cpp/src/neighbors/detail/ann_utils.cuh

```diff
@@ -471,12 +485,38 @@ struct batch_load_iterator {
                          size_t(offset()),
                          size_t(size()),
                          size_t(row_width()));
-          raft::copy(dev_ptr_, source_ + offset() * row_width(), size() * row_width(), stream_);
+          if (prefetch_ && prefetch_pos_ == pos_) {
+            std::swap(dev_ptr_, prefetch_dev_ptr_);
+          } else {
+            raft::copy(dev_ptr_, source_ + offset() * row_width(), size() * row_width(), stream_);
+          }
         }
       } else {
         dev_ptr_ = const_cast<T*>(source_) + offset() * row_width();
       }
     }
+
+    /**
+     * Helper function for prefetch. NOP if prefetch option is not enabled.
+     */
+    void prefetch(const size_type& pos)
+    {
+      if (pos >= n_iters_ || !prefetch_ || !needs_copy_ || source_ == nullptr) { return; }
+      size_type prefetch_offset = batch_size_ * pos;
+      size_type prefetch_size = std::min(batch_size_, n_rows_ - std::min(prefetch_offset, n_rows_));
+      raft::common::nvtx::push_range(
+        "batch_load_iterator::prefetch(offset = %zu, size = %zu, row_width = %zu)",
+        size_t(prefetch_offset),
+        size_t(prefetch_size),
+        size_t(row_width()));
+      raft::copy(prefetch_dev_ptr_,
+                 source_ + prefetch_offset * row_width(),
+                 prefetch_size * row_width(),
+                 stream_);
+      raft::common::nvtx::pop_range();
+      stream_.synchronize();
```

Comment:
I am wondering whether this is the most intuitive place to sync. Before looking into the implementation details, I would have expected, that prefetch, like most of our memory operations are asynchronous, but it is apparently not. If we want to keep this, then please mention it in the docstring. Alternatively, one could consider syncing before load swaps pointers, or we could leave it up to the user. Any thoughts @achirkin?

Reply:
One thing to keep in mind that we are currently using pageable buffers, so the copy is synchronous anyways, the extra sync call just makes that more explicit.

Reply:
Pageable buffer copies is not entirely sync -- see [here, Synchronous point 1](https://docs.nvidia.com/cuda/cuda-driver-api/api-sync-behavior.html), the copy is sync until the copy to staging buffer is done. 

For this particular case, I think we could leave it to users. Users could add stream record/event wait APIs to ensure the pipeline is correct. What do you think?

Reply:
We decided to keep the API as it is and add proper documentation.